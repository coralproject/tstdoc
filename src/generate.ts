import fs from "fs";
import path from "path";
import ts from "typescript";
import * as tjs from "typescript-json-schema";
import RefParser from "@apidevtools/json-schema-ref-parser";

import template from "./template";

/**
 * findCompilerOptions will find the compiler options for.
 *
 * @param fileName path to the file that we want to find the tsconfig.json for
 */
function findCompilerOptions(fileName: string) {
  // Find tsconfig file.
  const file = ts.findConfigFile(fileName, fs.existsSync);
  if (!file) {
    throw new Error(`tsconfig.json file not found for: ${fileName}`);
  }

  const text = fs.readFileSync(file).toString();
  const result = ts.parseConfigFileTextToJson(file, text);
  if (result.error) {
    throw result.error;
  }

  // Parse the JSON raw data into actual consumable compiler options.
  const config = ts.parseJsonConfigFileContent(
    result.config,
    ts.sys,
    path.dirname(file)
  );

  return config.options;
}

interface Options {
  /**
   * indent is the number of header characters to start each header with
   * (minimum 1).
   */
  indent: number;

  /**
   * fileName is the file that contains the typescript type that you want to
   * generate documentation for.
   */
  fileName: string;

  /**
   * symbolNames are the symbols in the file that you want to generate
   * documentation for.
   */
  symbolNames: string[];
}

interface Entry {
  symbolName: string;
  definition: tjs.Definition;
}

async function generate(options: Options) {
  const compilerOptions = findCompilerOptions(options.fileName);

  // Generate the program based on the input files.
  const program = tjs.getProgramFromFiles([options.fileName], compilerOptions);

  // Create the generator to generate the types.
  const generator = tjs.buildGenerator(program, { ignoreErrors: true });
  if (!generator) {
    throw new Error("could not create the generator");
  }

  // Get the schema for each symbol.
  const definitions: Entry[] = [];
  for (const symbolName of options.symbolNames) {
    // Get the JSON schema for the selected symbol.
    let schema = generator.getSchemaForSymbol(symbolName, true);

    // Dereference the references from this.
    let definition = await RefParser.dereference(
      schema as RefParser.JSONSchema
    );

    // Put the reference into a list.
    definitions.push({ symbolName, definition: definition as tjs.Definition });
  }

  // Setup the render context.
  const context = {
    // Provide the definitions.
    definitions,
    // Provide the default indent level for headings.
    indent: "#".repeat(options.indent),
  };

  // Render the template with the provided context.
  return template.render(context);
}

export default generate;
