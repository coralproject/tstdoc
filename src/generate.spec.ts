import generate from "./generate";
import path from "path";

export enum ROLE {
  COMMENTER = "COMMENTER",
  STAFF = "STAFF",
  MODERATOR = "MODERATOR",
  ADMIN = "ADMIN",
}

export interface TestInterfaceWithOptions {
  /**
   * action refers to the specific operation being performed. If `NEW`, this
   * is referring to a new comment being created. If `EDIT`, then this refers to
   * an operation involving an edit operation on an existing Comment.
   */
  action: "NEW" | "EDIT";

  /**
   * comment refers to the actual Comment data for the Comment being
   * created/edited.
   */
  comment: {
    /**
     * body refers to the actual body text of the Comment being created/edited.
     */
    body: string;

    /**
     * parentID is the identifier for the parent comment (if this Comment is a
     * reply, null otherwise).
     */
    parentID: string | null;
  };

  /**
   * author refers to the User that is creating/editing the Comment.
   */
  author: {
    /**
     * id is the identifier for this User.
     */
    id: string;

    /**
     * role refers to the role of this User.
     */
    role: ROLE;
  };

  /**
   * story refers to the Story being commented on.
   */
  story: {
    /**
     * id is the identifier for this Story.
     */
    id: string;

    /**
     * url is the URL for this Story.
     */
    url: string;
  };

  /**
   * site refers to the Site that the story being commented on belongs to.
   */
  site: {
    /**
     * id is the identifier for this Site.
     */
    id: string;
  };

  /**
   * tenantID is the identifer of the Tenant that this Comment is being
   * created/edited on.
   */
  tenantID: string;

  /**
   * tenantDomain is the domain that is associated with this Tenant that this
   * Comment is being created/edited on.
   */
  tenantDomain: string;
}

it("generates the expected form", async () => {
  const documentation = await generate({
    indent: 3,
    fileName: path.join(__dirname, "generate.spec.ts"),
    symbolNames: ["TestInterfaceWithOptions"],
  });

  expect(documentation).toMatchSnapshot();
});
