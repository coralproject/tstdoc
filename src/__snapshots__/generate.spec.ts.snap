// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates the expected form 1`] = `
"
### Parameters

#### \`action\` _(string, enum)_

action refers to the specific operation being performed. If \`NEW\`, this is referring to a new comment being created. If \`EDIT\`, then this refers to an operation involving an edit operation on an existing Comment.

Possible enum values for \`action\`:

- \`EDIT\`
- \`NEW\`

#### \`author\` _(object)_

author refers to the User that is creating/editing the Comment.

##### Attributes

- \`author.id\` _(string)_ - id is the identifier for this User.
- \`author.role\` _(string, enum)_ - role refers to the role of this User. Possible enum values:
  - \`ADMIN\`
  - \`COMMENTER\`
  - \`MODERATOR\`
  - \`STAFF\`


#### \`comment\` _(object)_

comment refers to the actual Comment data for the Comment being created/edited.

##### Attributes

- \`comment.body\` _(string)_ - body refers to the actual body text of the Comment being created/edited.
- \`comment.parentID\` _(null,string)_ - parentID is the identifier for the parent comment (if this Comment is a reply, null otherwise).

#### \`site\` _(object)_

site refers to the Site that the story being commented on belongs to.

##### Attributes

- \`site.id\` _(string)_ - id is the identifier for this Site.

#### \`story\` _(object)_

story refers to the Story being commented on.

##### Attributes

- \`story.id\` _(string)_ - id is the identifier for this Story.
- \`story.url\` _(string)_ - url is the URL for this Story.

#### \`tenantDomain\` _(string)_

tenantDomain is the domain that is associated with this Tenant that this Comment is being created/edited on.


#### \`tenantID\` _(string)_

tenantID is the identifer of the Tenant that this Comment is being created/edited on.


"
`;
